core/domain/prod_validation_jobs_one_off.py

```diff
@@ -58,22 +100,37 @@ def _validate_external_id_relationships(cls, item):
                 multiple_models_keys_to_fetch[field_name] = (
                     model_class, [field_value] if isinstance(field_value, str)
                     else field_value)
+        for field_name_debug, model_class, keys_to_fetch in (
+                cls._custom_external_id_relationship(item)):
+            multiple_models_keys_to_fetch[field_name_debug] = (
+                model_class, keys_to_fetch)
         fetched_model_instances = (
             datastore_services.fetch_multiple_entities_by_ids_and_models(
                 multiple_models_keys_to_fetch.values()))
-
-        for (field_name, keys_to_fetch), model_class_models in zip(
+        for (field_name, (model_class, field_values)), external_models in zip(
                 multiple_models_keys_to_fetch.iteritems(),
                 fetched_model_instances):
-            model_class, model_ids = keys_to_fetch
-            for model_id, model in zip(model_ids, model_class_models):
-                if model is None or model.deleted:
-                    cls.errors['%s field check' % field_name] = (
-                        'Model id %s: based on field %s having'
-                        ' value %s, expect model %s with id %s but it doesn\'t'
-                        ' exist' % (
-                            item.id, field_name, model_id,
-                            str(model_class.__name__), model_id))
+            cls.external_models[field_name] = (
+                model_class, zip(field_values, external_models))
+
+    @classmethod
+    def validate(cls, item):
+        """Run _fetch_external_models and all _validate functions.
+
+        The order of running _validate is not guaranteed.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+        """
+        cls.errors.clear()
+        cls.external_models.clear()
+        cls._fetch_external_models(item)
+
+        validate_fns = (
```

Comment:
I think this merits some documentation at the top of the file to explain that, for each class, all _validate* functions are run in random order, and that just writing such a function is enough.

Also, have you considered the alternative of doing this explicitly instead?

    cls._validate_external_id_relationships()
    cls._validate()

cls._validate() could contain multiple "blocks", one after the other, with a line comment above each block. (If desired, it could call other internal methods.) That makes this a bit more deterministic and would guard against accidental and easily-overlooked errors like forgetting the leading underscore.

Reply:
This makes sense and I'm okay with the explicit approach.